docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+
+#### `DataFlowImplSpecific.qll`
+```ql
+module Private {
+  import DataFlowPrivate
+}
+
+module Public {
+  import DataFlowPublic
+}
+```
+
+## Defining the data-flow graph
+
+The main input to the library is the data-flow graph. One must define a class
+`Node` and an edge relation `simpleLocalFlowStep(Node node1, Node node2)`. The
+`Node` class should be in `DataFlowPublic`.
+
+Recommendations:
+* Make `Node` an IPA type. There is commonly a need for defining various
+  data-flow nodes that are not necessarily represented in the AST of the
+  language.
+* Define `predicate localFlowStep(Node node1, Node node2)` as an alias of
+  `simpleLocalFlowStep` and expose it publicly. The reason for this indirection
+  is that it gives the option of exposing local flow augmented with field flow.
+  See the C/C++ implementation, which makes use of this feature.
+* Define `predicate localFlow(Node node1, Node node2) { localFlowStep*(node1, node2) }`.
+* Make the local flow step relation in `simpleLocalFlowStep` follow
+  def-to-first-use and use-to-next-use steps for SSA variables. Def-use steps
+  also work, but the upside of `use-use` steps is that sources defined in terms
+  of variable reads just work out of the box. It also makes certain
+  barrier-implementations simpler.
+
+The shared library does not use `localFlowStep` nor `localFlow` but users of
+`DataFlow.qll` may expect the existence of `DataFlow::localFlowStep` and
+`DataFlow::localFlow`.
+
+### `Node` subclasses
+
+The `Node` class needs a number of subclasses. As a minimum the following are needed:
+```
+ExprNode
+ParameterNode
+PostUpdateNode
+
+OutNode
+ArgumentNode
+ReturnNode
+CastNode
+```
+and possibly more depending on the language and its AST. Of the above, the
+first 3 should be public, but the last 4 can be private. Also, the last 4 will
+likely be subtypes of `ExprNode`. For further details about `ParameterNode`,
+`ArgumentNode`, `ReturnNode`, and `OutNode` see [The call-graph](#the-call-graph)
+below. For further details about `CastNode` see [Type pruning](#type-pruning) below.
+For further details about `PostUpdateNode` see [Field flow](#field-flow) below.
+
+Nodes corresponding to expressions and parameters are the most common for users
+to interact with so a couple of convenience predicates are generally included:
+```
+DataFlowExpr Node::asExpr()
+Parameter Node::asParameter()
+ExprNode exprNode(DataFlowExpr n)
+ParameterNode parameterNode(Parameter n)
+```
+Here `DataFlowExpr` should be an alias for the language-specific class of
+expressions (typically called `Expr`). Parameters do not need an alias for the
+shared implementation to refer to, so here you can just use the
+language-specific class name (typically called `Parameter`).
+
+### The call-graph
+
+In order to make inter-procedural flow work a number of classes and predicates
+must be provided.
+
+First, two types, `DataFlowCall` and `DataFlowCallable`, must be defined. These
```

Comment:
Perhaps mention that it may also make sense to use IPA types (for C#, `DataFlowCall` is [an IPA type](https://github.com/github/codeql/blob/master/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll#L89), which enables us to model for example implicit calls to delegates that are passed as arguments to library code).